internal/service/creditgrant.go

```diff
@@ -219,3 +221,546 @@ func (s *creditGrantService) GetCreditGrantsBySubscription(ctx context.Context,
+	for _, cga := range applications {
+		// Skip if already applied
+		if cga.ApplicationStatus == types.ApplicationStatusApplied {
+			continue
+		}
+
+		// Set tenant and environment context
+		ctxWithTenant := context.WithValue(ctx, types.CtxTenantID, cga.TenantID)
+		ctxWithEnv := context.WithValue(ctxWithTenant, types.CtxEnvironmentID, cga.EnvironmentID)
+
+		err := s.processScheduledApplication(ctxWithEnv, cga)
+		if err != nil {
+			s.Logger.Errorw("failed to process scheduled application",
+				"application_id", cga.ID,
+				"grant_id", cga.CreditGrantID,
+				"subscription_id", cga.SubscriptionID,
+				"error", err)
+		}
+	}
+
+	return nil
+}
+
+// processScheduledApplication processes a single scheduled credit grant application
+func (s *creditGrantService) processScheduledApplication(
+	ctx context.Context,
+	cga *domainCreditGrantApplication.CreditGrantApplication,
+) error {
+	subscriptionService := NewSubscriptionService(s.ServiceParams)
+	creditGrantService := NewCreditGrantService(s.ServiceParams)
+
+	// Get subscription
+	subscription, err := subscriptionService.GetSubscription(ctx, cga.SubscriptionID)
+	if err != nil {
+		s.Logger.Errorw("failed to get subscription", "subscription_id", cga.SubscriptionID, "error", err)
+		return err
+	}
+
+	// Get credit grant
+	creditGrant, err := creditGrantService.GetCreditGrant(ctx, cga.CreditGrantID)
+	if err != nil {
+		s.Logger.Errorw("failed to get credit grant", "credit_grant_id", cga.CreditGrantID, "error", err)
+		return err
+	}
+
+	// Check if credit grant is published
+	if creditGrant.CreditGrant.Status != types.StatusPublished {
+		s.Logger.Debugw("credit grant is not published, skipping", "credit_grant_id", cga.CreditGrantID)
+		return nil
+	}
+
+	// If exists and applied successfully, skip
+	if cga.ApplicationStatus == types.ApplicationStatusApplied {
+		s.Logger.Debugw("grant already applied successfully, skipping", "application_id", cga.ID)
+		return nil
+	}
+
+	// If exists and failed, retry
+	if cga.ApplicationStatus == types.ApplicationStatusFailed {
+		return s.retryFailedApplication(ctx, cga, creditGrant.CreditGrant, subscription.Subscription)
+	}
+
+	// check if the credit grant is already applied for this period using proper CGA tracking
+	existingApp, err := s.CheckDuplicateApplication(ctx, creditGrant.CreditGrant, subscription.Subscription, cga.PeriodStart, cga.PeriodEnd)
+	if err != nil {
+		s.Logger.Errorw("failed to check for duplicate application", "error", err)
+		return err
+	}
+
+	if existingApp != nil && existingApp.ApplicationStatus == types.ApplicationStatusApplied {
+		s.Logger.Debugw("grant already applied for this period, skipping",
+			"application_id", cga.ID,
+			"existing_application_id", existingApp.ID)
+		return nil
+	}
+
+	// Apply the grant
+	return s.applyScheduledGrant(ctx, creditGrant.CreditGrant, subscription.Subscription, cga)
+}
+
+// applyScheduledGrant applies a scheduled credit grant
+func (s *creditGrantService) applyScheduledGrant(
```

Comment:
no need of this move the block to 
```
stateHandler := NewSubscriptionStateHandler(subscription, grant)
action, reason := stateHandler.DetermineAction()
```

move to processScheduledApplication 

And use the common function